### PR TITLE
Require acknowledgement to start_trial license

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This is related to #30134. It modifies the start_trial action to require
an acknowledgement parameter in the rest request to actually start the
trial license. There are backwards compatibility issues as prior ES
versions did not support this parameter. To handle this, it is assumed
that a request coming from a node prior to 6.3 is acknowledged. And
attempts to write a non-acknowledged request to a prior to 6.3 node will
throw an exception.

Additionally this PR adds messages about the trial license the user is
generating.

### CODE REVIEW

**Comment by tbrooks8** [comment]
Here is the message for a trial that needs to be acknowledge:

{
"acknowledged": false,
"trial_was_started": false,
"error_message": "Operation failed: Needs acknowledgement.",
"acknowledge": {
"message": "This API initiates a free 30-day trial for all platinum features. By starting this trial, you agree that it is subject to the terms and conditions at https://www.elastic.co/legal/trial_license/. To begin your free trial, call /start_trial again and specify the "acknowledge=true" parameter.",
"security": [
"With a trial license, X-Pack security features are available, but are not enabled by default."
]
}
}
Trial that is acknowledged:

{
"acknowledged": true,
"trial_was_started": true,
"type": "trial"
}

File: x-pack/docs/build.gradle.orig [hunk 1]
```diff
@@ -0,0 +1,687 @@
+import org.elasticsearch.gradle.test.NodeInfo
```
Comment (jaymode) [inline x-pack/docs/build.gradle.orig]: remove this file

File: x-pack/docs/en/rest-api/license/start-trial.asciidoc [hunk 2]
```diff
@@ -40,7 +40,7 @@ The following example checks whether you are eligible to start a trial:
 
 [source,js]
 ------------------------------------------------------------
-POST _xpack/license/start_trial
+GET _xpack/license/start_trial
```
Comment (jaymode) [inline x-pack/docs/en/rest-api/license/start-trial.asciidoc]: I think we should make `start_trial` start with an underscore like other APIs that perform an action (`_search`, `_authenticate`, etc), so `_start_trial` and `_start_basic`

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/LicensingClient.java.orig [hunk 3]
```diff
@@ -0,0 +1,67 @@
+/*
```
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/LicensingClient.java.orig]: remove this file

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/PostStartTrialRequest.java.orig [hunk 4]
```diff
@@ -0,0 +1,55 @@
+/*
```
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/PostStartTrialRequest.java.orig]: remove this file

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/PostStartTrialRequestBuilder.java.orig [hunk 5]
```diff
@@ -0,0 +1,17 @@
+/*
```
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/PostStartTrialRequestBuilder.java.orig]: remove this file

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/PostStartTrialResponse.java.orig [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/*
```
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/PostStartTrialResponse.java.orig]: remove this file

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/RestGetTrialStatus.java.orig [hunk 7]
```diff
@@ -0,0 +1,48 @@
+/*
```
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/RestGetTrialStatus.java.orig]: remove this file

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/RestPostStartTrialLicense.java.orig [hunk 8]
```diff
@@ -0,0 +1,59 @@
+/*
```
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/RestPostStartTrialLicense.java.orig]: remove this file

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/StartTrialClusterTask.java.orig [hunk 9]
```diff
@@ -0,0 +1,79 @@
+/*
```
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/StartTrialClusterTask.java.orig]: remove this file

File: x-pack/plugin/core/src/test/java/org/elasticsearch/license/StartTrialLicenseTests.java.orig [hunk 10]
```diff
@@ -0,0 +1,113 @@
+/*
```
Comment (jaymode) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/license/StartTrialLicenseTests.java.orig]: remove this file

File: x-pack/plugin/src/test/resources/rest-api-spec/test/license/20_put_license.yml [hunk 11]
```diff
@@ -133,7 +133,8 @@ teardown:
 
   - do:
       catch: forbidden
-      xpack.license.post_start_trial: {}
+      xpack.license.post_start_trial:
```
Comment (jaymode) [inline x-pack/plugin/src/test/resources/rest-api-spec/test/license/20_put_license.yml]: should it still be post_start_trial and not changed to get_start_trial? Or just start_trial

**Review by jaymode — CHANGES_REQUESTED** [review]

File: x-pack/plugin/src/test/resources/rest-api-spec/test/license/20_put_license.yml (hunk 11 above)
Comment (tbrooks8) [reply x-pack/plugin/src/test/resources/rest-api-spec/test/license/20_put_license.yml]: No it should be post_start_trial. We're testing that this call fails.

File: x-pack/docs/en/rest-api/license/start-trial.asciidoc (hunk 2 above)
Comment (tbrooks8) [reply x-pack/docs/en/rest-api/license/start-trial.asciidoc]: https://github.com/elastic/x-pack-elasticsearch/issues/1941#issuecomment-322642475

That was discussed. Also I imagine that this "bug fix" PR to add necessary ack messages to a response would not be the place to modify a route that already exists in prior versions of ES x-pack.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/PostStartTrialResponse.java [hunk 12]
```diff
@@ -57,10 +76,57 @@ public Status getStatus() {
     @Override
     public void readFrom(StreamInput in) throws IOException {
         status = in.readEnum(Status.class);
+        if (in.getVersion().onOrAfter(Version.V_6_3_0)) {
```
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/PostStartTrialResponse.java]: the readFrom version is different than the writeTo. Doesn't seem correct to me

**Review by jaymode — APPROVED** [review]
Left one comment, but otherwise LGTM

**Comment by tbrooks8** [comment]
run sample packaging tests

**Comment by tbrooks8** [comment]
run gradle build tests

**Pull request merged**